kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/internal/ProxyChannelStateMachineTest.java

```diff
@@ -70,18 +77,131 @@
 @TestInstance(TestInstance.Lifecycle.PER_CLASS)
 class ProxyChannelStateMachineTest {
 
-    public static final HostPort BROKER_ADDRESS = new HostPort("localhost", 9092);
-    public static final HAProxyMessage HA_PROXY_MESSAGE = new HAProxyMessage(HAProxyProtocolVersion.V2, HAProxyCommand.PROXY, HAProxyProxiedProtocol.TCP4,
+    private static final HostPort BROKER_ADDRESS = new HostPort("localhost", 9092);
+    private static final HAProxyMessage HA_PROXY_MESSAGE = new HAProxyMessage(HAProxyProtocolVersion.V2, HAProxyCommand.PROXY, HAProxyProxiedProtocol.TCP4,
             "1.1.1.1", "2.2.2.2", 46421, 9092);
+    private static final Offset<Double> BLOODY_CLOSE = Offset.offset(0.00005);
+    private static final String CLUSTER_NAME = "virtualClusterA";
+    private static final VirtualClusterModel VIRTUAL_CLUSTER_MODEL = new VirtualClusterModel(CLUSTER_NAME, new TargetCluster("", Optional.empty()), false, false,
+            List.of());
+    private final RuntimeException failure = new RuntimeException("There's Klingons on the starboard bow");
     private ProxyChannelStateMachine proxyChannelStateMachine;
     private KafkaProxyBackendHandler backendHandler;
     private KafkaProxyFrontendHandler frontendHandler;
+    private SimpleMeterRegistry simpleMeterRegistry;
 
     @BeforeEach
     void setUp() {
-        proxyChannelStateMachine = new ProxyChannelStateMachine();
+        proxyChannelStateMachine = new ProxyChannelStateMachine(CLUSTER_NAME);
         backendHandler = mock(KafkaProxyBackendHandler.class);
         frontendHandler = mock(KafkaProxyFrontendHandler.class);
+        simpleMeterRegistry = new SimpleMeterRegistry();
+        Metrics.globalRegistry.add(simpleMeterRegistry);
+    }
+
+    @AfterEach
+    void tearDown() {
+        if (simpleMeterRegistry != null) {
+            simpleMeterRegistry.getMeters().forEach(Metrics.globalRegistry::remove);
+            Metrics.globalRegistry.remove(simpleMeterRegistry);
+        }
+    }
+
+    @Test
+    void shouldCountClientConnections() {
+        // Given
+
+        // When
+        proxyChannelStateMachine.onClientActive(frontendHandler);
+
+        // Then
+        assertThat(Metrics.globalRegistry.get("kroxylicious_downstream_connections").counter())
+                .isNotNull()
+                .satisfies(counter -> assertThat(counter.getId()).isNotNull())
+                .satisfies(counter -> assertThat(counter.count())
+                        .isCloseTo(1.0, BLOODY_CLOSE));
+    }
+
+    @ParameterizedTest
+    @MethodSource("clientErrorStates")
+    void shouldCountClientExceptions(Runnable givenState, Boolean tlsEnabled) {
+        // Given
+        givenState.run();
+
+        // When
+        proxyChannelStateMachine.onClientException(failure, tlsEnabled);
+
+        // Then
+        assertThat(Metrics.globalRegistry.get("kroxylicious_downstream_errors").counter())
```

Comment:
wonder if we should look into rolling some assertj code in future so we can do something like:

```
assertThatGlobalMetric("kroxylicious_downstream_errors")
                .isNotNull()
                .isCounter()
                .hasCountCloseTo(1.0, BLOODY_CLOSE);
```

Reply:
`CLOSE_ENOUGH` :-)

Reply:
Yeah thats an interesting idea for another day/PR.

Reply:
`WITHIN_A_NATS_WHISKER` 😜 

The current constant is probably not right, but I wanted a bit of humour.